File: android/src/main/kotlin/com/benjamin/horner/helpers/HexHelper.kt
package com.benjamin.horner.flutter_acs_card_reader

import com.benjamin.horner.flutter_acs_card_reader.StringHelper
import java.nio.charset.StandardCharsets
import android.util.Log

class HexHelper {
    private val TAG = "HexHelper"
    private val stringHelper = StringHelper()

    fun byteArrayToHexString(buffer: ByteArray): String {
        var bufferString = ""
        for (i in buffer.indices) {
            var hexChar = (buffer[i].toInt() and 0xFF).toString(16)
                .padStart(2, '0') // Ensure two-character representation
    
            bufferString += hexChar.toUpperCase() + " "
        }
        
        return bufferString.trim() // Remove trailing space, if any
    }
    
    fun hexStringToByteArray(hexString: String): ByteArray {
        val sanitized = hexString.replace(" ", "") // Remove spaces if present
        val length = sanitized.length
        val byteArray = ByteArray(length / 2)
    
        for (i in 0 until length step 2) {
            val byte = sanitized.substring(i, i + 2).toInt(16).toByte()
            byteArray[i / 2] = byte
        }
    
        return byteArray
    }

    fun cleanupHexString(hexString: String): String {
        return stringHelper.removeWhitespaces(hexString)
    }

    fun byteLength(apdu: ApduCommand? = null, length: Int = 0): String {
        if (apdu == null && length > 0) {
            return padHex(Integer.toHexString(length)).toUpperCase()
        } else if (apdu!!.lengthMin == apdu!!.lengthMax && apdu!!.lengthMax <= 255) {
            return padHex(Integer.toHexString(apdu!!.lengthMin)).toUpperCase()
        } else if (length > 0) {
            return padHex(Integer.toHexString(length)).toUpperCase()
        } else {
            return padHex(Integer.toHexString(255)).toUpperCase()
        }
    }

    fun calculateLengthOfHex(hexString: String): String {
        val length: Int = cleanupHexString(hexString).length/2
        return padHex(hexString = Integer.toHexString(length), desiredLength = 4).toUpperCase()
    }

    fun calculateLengthToHex(length: Int): String {
        return padHex(hexString = Integer.toHexString(length), desiredLength = 4).toUpperCase()
    }

    fun padHex(hexString: String, desiredLength: Int = 2): String {
        val paddedHex = String.format("%${desiredLength}s", hexString).replace(' ', '0')

        val spacedHex = buildString {
            for (i in 0 until paddedHex.length step 2) {
                if (i > 0) {
                    append(' ') // Add a space after every 2 characters
                }
                append(paddedHex.substring(i, i + 2))
            }
        }

        return spacedHex.toUpperCase()
    }

    // private fun padHex(hex: String): String {
    //     val paddedHex = if (hex.length % 2 == 1) {
    //         "0$hex"
    //     } else {
    //         hex
    //     }

    //     val spacedHex = buildString {
    //         for (i in 0 until paddedHex.length step 2) {
    //             if (i > 0) {
    //                 append(' ') // Add a space after every 2 characters
    //             }
    //             append(paddedHex.substring(i, i + 2))
    //         }
    //     }

    //     return spacedHex
    // }
 
}

File: ios/Classes/Helpers/HexHelper.swift
//
//  HexHelper.swift
//  flutter_acs_card_reader
//
//  Created by Benjamin Horner on 23/11/2023.
//

import Foundation

class HexHelper {
    private let TAG = "HexHelper"
    private let stringHelper = StringHelper()

    func byteArrayToHexString(buffer: [UInt8]) -> String {
        var bufferString = ""
        for byte in buffer {
            let hexChar = String(format: "%02X", byte)
            bufferString += hexChar + " "
        }
        return bufferString.trimmingCharacters(in: .whitespaces)
    }

    func hexStringToByteArray(hexString: String) -> [UInt8] {
        let sanitized = hexString.replacingOccurrences(of: " ", with: "")
        let length = sanitized.count
        var byteArray = [UInt8]()

        for i in stride(from: 0, to: length, by: 2) {
            let startIndex = sanitized.index(sanitized.startIndex, offsetBy: i)
            let endIndex = sanitized.index(startIndex, offsetBy: 2)
            let byteString = sanitized[startIndex..<endIndex]
            if let byte = UInt8(byteString, radix: 16) {
                byteArray.append(byte)
            }
        }

        return byteArray
    }

    func cleanupHexString(hexString: String) -> String {
        return stringHelper.removeWhitespaces(input: hexString)
    }

    func byteLength(apdu: ApduCommand? = nil, length: Int = 0) -> String {
        if let apdu = apdu, apdu.lengthMin == apdu.lengthMax, apdu.lengthMax <= 255 {
            return padHex(hexString: String(apdu.lengthMin, radix: 16), desiredLength: 2).uppercased()
        } else {
            return padHex(hexString: String(length, radix: 16), desiredLength: 2).uppercased()
        }
    }

    func calculateLengthOfHex(hexString: String) -> String {
        let length = cleanupHexString(hexString: hexString).count / 2
        return padHex(hexString: String(format: "%04X", length), desiredLength: 4).uppercased()
    }

    func calculateLengthToHex(length: Int) -> String {
        return padHex(hexString: String(format: "%04X", length), desiredLength: 4).uppercased()
    }

    func padHex(hexString: String, desiredLength: Int = 2) -> String {
        let paddedHex = String(repeating: "0", count: max(0, desiredLength - hexString.count)) + hexString

        let spacedHex = stride(from: 0, to: paddedHex.count, by: 2).map {
            paddedHex[paddedHex.index(paddedHex.startIndex, offsetBy: $0)..<paddedHex.index(paddedHex.startIndex, offsetBy: $0 + 2)]
        }.joined(separator: " ")

        return spacedHex.uppercased()
    }
}

